### PR TITLE
Honor icu locale variant

### PR DESCRIPTION
**Pull request by alexbakharew**
Added support for variant in locale. Added new tests

### CODE REVIEW

File: core/analysis/collation_token_stream.cpp [hunk 1]
```diff
@@ -37,7 +37,119 @@ namespace {
...
 using namespace irs;
 
-constexpr VPackStringRef LOCALE_PARAM_NAME{"locale"};
+constexpr VPackStringRef LOCALE_PARAM_NAME    {"locale"};
+constexpr VPackStringRef LANGUAGE_PARAM_NAME  {"language"};
+constexpr VPackStringRef COUNTRY_PARAM_NAME   {"country"};
+constexpr VPackStringRef ENCODING_PARAM_NAME  {"encoding"};
+constexpr VPackStringRef VARIANT_PARAM_NAME   {"variant"};
+
+std::string get_locale_name(const string_ref language,
```
Comment (Dronplane) [inline core/analysis/collation_token_stream.cpp:46]: Maybe move this function to locale_utils.hpp/cpp ?

File: core/analysis/collation_token_stream.cpp [hunk 2]
```diff
@@ -37,7 +37,119 @@ namespace {
...
+    variant = get_string<string_ref>(variant_slice);
+  }
+
+  auto loc_name = get_locale_name(language, country, encoding, variant);
+  locale = locale_utils::locale(loc_name);
+
+  return true;
+}
+
+bool verify_icu_locale(const std::locale& locale) {
```
Comment (Dronplane) [inline core/analysis/collation_token_stream.cpp:124]: Maybe move this function to locale_utils.hpp/cpp ?

File: core/analysis/collation_token_stream.cpp [hunk 3]
```diff
@@ -112,8 +226,21 @@ bool make_vpack_config(
   VPackObjectBuilder object(builder);
   {
     // locale
-    const auto& locale_name = locale_utils::name(options.locale);
-    builder->add(LOCALE_PARAM_NAME, VPackValue(locale_name));
+    {
```
Comment (Dronplane) [inline core/analysis/collation_token_stream.cpp:229]: As general suggestion:  Let's move all locale object serializing / deserializing to the separate compilation unit. So we could use it later for other analyzers (like text, norm and so on)

File: tests/analysis/collation_token_stream_test.cpp [hunk 4]
```diff
@@ -58,6 +58,98 @@ TEST(collation_token_stream_test, construct) {
   }
 }
 
+TEST(collation_token_stream_test, check_variant) {
```
Comment (Dronplane) [inline tests/analysis/collation_token_stream_test.cpp:61]: Please add tests for parsing different locale objects including invalid ones. 
(E.g.   R"({"locale": {"language" : "de",
                     "encoding" : true,
                     "variant" :  123232}}) )

**Review by Dronplane — CHANGES_REQUESTED** [review]

File: core/utils/icu_locale_utils.cpp [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+  }
+
+  string_ref language;
+  string_ref country;
+  string_ref variant;
+
+  auto lang_slice = slice.get(LANGUAGE_PARAM_NAME);
+  if (lang_slice.isNone() || !lang_slice.isString()) {
+    IR_FRMT_WARN(
+      "Language parameter is not specified or it is not a string",
```
Comment (Dronplane) [inline core/utils/icu_locale_utils.cpp:51]: Looks like we are missing '%s' format specifier somewhere

File: core/utils/icu_locale_utils.cpp [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+      LANGUAGE_PARAM_NAME.data());
+    return false;
+  }
+  language = get_string<string_ref>(lang_slice);
+
+  if (slice.hasKey(COUNTRY_PARAM_NAME)) {
+    auto country_slice = slice.get(COUNTRY_PARAM_NAME);
+    if (!country_slice.isString()) {
+      return false;
```
Comment (Dronplane) [inline core/utils/icu_locale_utils.cpp:60]: Should we write something to log in case of invalid type ?

File: core/utils/icu_locale_utils.cpp [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    }
+    country = get_string<string_ref>(country_slice);
+  }
+
+  if (slice.hasKey(VARIANT_PARAM_NAME)) {
+    auto variant_slice = slice.get(VARIANT_PARAM_NAME);
+    if (!variant_slice.isString()) {
+      return false;
```
Comment (Dronplane) [inline core/utils/icu_locale_utils.cpp:68]: Should we write something to log in case of invalid type ?

File: core/utils/icu_locale_utils.cpp [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+    locale_name.append(variant.c_str(), variant.size());
+  }
+
+  locale = icu::Locale::createFromName(locale_name.c_str());
+
+  return true;
+}
+
+bool locale_to_vpack(const icu::Locale& locale, VPackBuilder* const builder) {
+  VPackObjectBuilder object(builder);
```
Comment (Dronplane) [inline core/utils/icu_locale_utils.cpp:88]: Do we really want this additional wrapper object? With this wrapper result of locale_to_vpack will not  be accepted by get_locale_from_vpack. that looks weird.

File: core/analysis/collation_token_stream.hpp [hunk 9]
```diff
@@ -26,6 +26,8 @@
 #include "analyzers.hpp"
 #include "token_attributes.hpp"
 #include "utils/frozen_attributes.hpp"
+#include "unicode/locid.h"
```
Comment (gnusi) [inline core/analysis/collation_token_stream.hpp:29]: this header has to be placed before iresearch headers as per guidelines, also use <> instead of quotes

File: core/utils/icu_locale_utils.hpp [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Alexey Bakharew
+////////////////////////////////////////////////////////////////////////////////
+
+#ifndef ICU_LOCALE_UTILS_HPP
+#define ICU_LOCALE_UTILS_HPP
+
+#include <unicode/locid.h>
```
Comment (gnusi) [inline core/utils/icu_locale_utils.hpp:26]: please honor guildelines
<system headers>
<external library headers>
<our own headers>

File: tests/utils/icu_locale_utils_tests.cpp [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+
+#include "gtest/gtest.h"
+#include "utils/locale_utils.hpp"
+#include "utils/icu_locale_utils.hpp"
+#include "utils/misc.hpp"
+#include "velocypack/Parser.h"
+#include "velocypack/velocypack-aliases.h"
+
+namespace tests {
+  class IcuLocaleUtilsTestSuite: public ::testing::Test {
```
Comment (gnusi) [inline tests/utils/icu_locale_utils_tests.cpp:31]: why empty test class?

**Review by gnusi — CHANGES_REQUESTED** [review]

File: core/utils/icu_locale_utils.cpp [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+
+  if (locale_slice.hasKey(ENCODING_PARAM_NAME)) {
+    auto encoding_slice = locale_slice.get(ENCODING_PARAM_NAME);
+    if (!encoding_slice.isString()) {
+      IR_FRMT_ERROR(
+        "'%s' parameter name should be string", ENCODING_PARAM_NAME.data());
+      return false;
+    }
+    encoding = get_string<string_ref>(encoding_slice);
+    if (encoding != "utf-8") {
```
Comment (Dronplane) [inline core/utils/icu_locale_utils.cpp:89]: Suggest to make this check case insensitive. As "UTF-8" and "utf-8" are both valid in general. Maybe some "tolower" ?

File: tests/analysis/collation_token_stream_test.cpp [hunk 13]
```diff
@@ -20,41 +20,317 @@
...
+    ASSERT_EQ(nullptr, stream);
+  }
+
+  // vpack
+  {
+    std::string config = R"({ "locale": {"language" : "en", "country": "US", "variant" : "phonebook"}})";
+    auto in_vpack = VPackParser::fromJson(config.c_str(), config.size());
+    std::string in_str;
+    in_str.assign(in_vpack->slice().startAs<char>(), in_vpack->slice().byteSize());
+    std::string out_str;
```
Comment (Dronplane) [inline tests/analysis/collation_token_stream_test.cpp:69]: Unused?

File: tests/analysis/collation_token_stream_test.cpp [hunk 14]
```diff
@@ -20,41 +20,317 @@
...
+      R"({ "locale": {"country": "US"}})");
+    ASSERT_EQ(nullptr, stream);
+  }
+
+  // vpack
+  {
+    std::string config = R"({ "locale": {"language" : "en", "country": "US", "variant" : "phonebook"}})";
+    auto in_vpack = VPackParser::fromJson(config.c_str(), config.size());
+    std::string in_str;
+    in_str.assign(in_vpack->slice().startAs<char>(), in_vpack->slice().byteSize());
```
Comment (Dronplane) [inline tests/analysis/collation_token_stream_test.cpp:68]: irs::string_ref could be created the same way without intermediate string.

File: tests/analysis/collation_token_stream_test.cpp [hunk 15]
```diff
@@ -20,41 +20,317 @@
...
+    std::string out_str;
+    auto stream = irs::analysis::analyzers::get("collation", irs::type<irs::text_format::vpack>::get(), in_str);
+    ASSERT_NE(nullptr, stream);
+    ASSERT_EQ(irs::type<irs::analysis::collation_token_stream>::id(), stream->type());
+  }
+
```
Comment (Dronplane) [inline tests/analysis/collation_token_stream_test.cpp:74]: Please add test cases:
language  and variant but no country
only variant
only country

File: tests/analysis/collation_token_stream_test.cpp [hunk 16]
```diff
@@ -20,41 +20,317 @@
...
+  // invalid
   {
     ASSERT_EQ(nullptr, irs::analysis::analyzers::get("collation", irs::type<irs::text_format::json>::get(), irs::string_ref::NIL));
     ASSERT_EQ(nullptr, irs::analysis::analyzers::get("collation", irs::type<irs::text_format::json>::get(), "1"));
     ASSERT_EQ(nullptr, irs::analysis::analyzers::get("collation", irs::type<irs::text_format::json>::get(), "[]"));
     ASSERT_EQ(nullptr, irs::analysis::analyzers::get("collation", irs::type<irs::text_format::json>::get(), "{}"));
     ASSERT_EQ(nullptr, irs::analysis::analyzers::get("collation", irs::type<irs::text_format::json>::get(), "{\"locale\":1}"));
+    ASSERT_EQ(nullptr, irs::analysis::analyzers::get("collation", irs::type<irs::text_format::json>::get(), R"({ "locale" : {"language" : "de_DE", "variant" : "_phonebook", "encoding" : "utf-32"}})"));
```
Comment (Dronplane) [inline tests/analysis/collation_token_stream_test.cpp:82]: Please also include checks for invalid value types for language object members

File: core/utils/icu_locale_utils.cpp [hunk 17]
```diff
@@ -0,0 +1,193 @@
...
+    locale_name.assign(locale_str.begin(), at_pos);
+  }
+
+  // extract encoding
+  if (!dot_pos) {
+    *encoding = Unicode::UTF8; // encoding is not specified. Set to default value
+  } else {
+    std::string enc;
+    if (at_pos) {
+      enc.assign(dot_pos + 1, at_pos); // encoding is located between '.' and '@' symbols
```
Comment (Dronplane) [inline core/utils/icu_locale_utils.cpp:143]: what if `dot_pos` is after `at_pos` ?  like in string "ru_RU@let.s.break.it"

File: tests/utils/icu_locale_utils_tests.cpp [hunk 18]
```diff
@@ -0,0 +1,318 @@
...
+    auto vpack = VPackParser::fromJson(config.c_str(), config.size());
+    icu::Locale actual_locale;
+    ASSERT_TRUE(get_locale_from_vpack(vpack->slice(), actual_locale));
+
+    icu::Locale expected_locale("DE", NULL, NULL, "collation=phonebook");
+    ASSERT_EQ(actual_locale, expected_locale);
+  }
+}
+
+TEST(icu_locale_utils_test_suite, get_locale_from_str) {
```
Comment (Dronplane) [inline tests/utils/icu_locale_utils_tests.cpp:107]: Please add some tests  with totally malformed string to be sure we could safely handle such input (e.g. do not crash).
Like "@ru_De.utf-8" or ".@".

**Review by Dronplane — CHANGES_REQUESTED** [review]

File: tests/utils/icu_locale_utils_tests.cpp [hunk 19]
```diff
@@ -235,6 +235,112 @@ TEST(icu_locale_utils_test_suite, get_locale_from_str) {
...
+    irs::string_ref expected = "ru_RU";
+    ASSERT_EQ(expected, actual);
+  }
+
+  // try to break it
+  {
+    irs::string_ref locale_name = "@ru_De.utf-8";
+    icu::Locale locale;
+    Unicode unicode;
+    ASSERT_TRUE(get_locale_from_str(locale_name, locale, true, &unicode));
```
Comment (Dronplane) [inline tests/utils/icu_locale_utils_tests.cpp:268]: Please add test with the same input but "new" parsing mode. That's another corner case

File: core/utils/icu_locale_utils.hpp [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+enum class Unicode { UTF8, NON_UTF8 };
+
+bool get_locale_from_vpack(const VPackSlice slice,
+                           icu::Locale& locale,
+                           Unicode* const unicode = nullptr);
+
+bool get_locale_from_str(irs::string_ref locale_str,
+                         icu::Locale& locale,
+                         bool is_new_format,
+                         Unicode* encoding = nullptr,
```
Comment (Dronplane) [inline core/utils/icu_locale_utils.hpp:45]:
```suggestion
                         Unicode* unicode = nullptr,
```

File: core/utils/icu_locale_utils.hpp [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+                         std::string* = nullptr);
```
Comment (Dronplane) [inline core/utils/icu_locale_utils.hpp:46]:
```suggestion
                         std::string* encoding = nullptr);
```

File: core/utils/icu_locale_utils.hpp [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+#include "velocypack/velocypack-aliases.h"
+
+namespace iresearch {
+namespace icu_locale_utils {
+
+enum class Unicode { UTF8, NON_UTF8 };
+
+bool get_locale_from_vpack(const VPackSlice slice,
+                           icu::Locale& locale,
+                           Unicode* const unicode = nullptr);
```
Comment (Dronplane) [inline core/utils/icu_locale_utils.hpp:40]:
```suggestion
                           Unicode* unicode = nullptr);
```

**Review by Dronplane — APPROVED** [review]
LGTM
